File: app/src/main/java/me/echeung/moemoekyun/client/socket/Socket.java
package me.echeung.moemoekyun.client.socket;

import android.os.Handler;
import android.os.SystemClock;
import android.util.Log;

import com.squareup.moshi.Moshi;

import java.io.IOException;

import lombok.RequiredArgsConstructor;
import lombok.Setter;
import me.echeung.moemoekyun.client.RadioClient;
import me.echeung.moemoekyun.client.auth.AuthUtil;
import me.echeung.moemoekyun.client.socket.response.BaseResponse;
import me.echeung.moemoekyun.client.socket.response.ConnectResponse;
import me.echeung.moemoekyun.client.socket.response.UpdateResponse;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;

@RequiredArgsConstructor
public class Socket extends WebSocketListener {

    private static final String TAG = Socket.class.getSimpleName();

    private static final Moshi MOSHI = new Moshi.Builder().build();

    private static final String TRACK_UPDATE = "TRACK_UPDATE";
    private static final String TRACK_UPDATE_REQUEST = "TRACK_UPDATE_REQUEST";
    private static final String QUEUE_UPDATE = "QUEUE_UPDATE";

    private static final int RETRY_TIME_MIN = 250;
    private static final int RETRY_TIME_MAX = 4000;
    private int retryTime = RETRY_TIME_MIN;
    private boolean attemptingReconnect = false;

    private final OkHttpClient client;
    private final AuthUtil authUtil;

    private WebSocket socket;
    @Setter
    private Listener listener;

    private Handler heartbeatHandler = new Handler();
    private Runnable heartbeatTask;

    public void connect() {
        Log.d(TAG, "Connecting to socket...");

        if (socket != null) {
            disconnect();
        }

        final Request request = new Request.Builder().url(RadioClient.getLibrary().getSocketUrl()).build();
        socket = client.newWebSocket(request, this);
    }

    public void disconnect() {
        clearHeartbeat();

        if (socket != null) {
            socket.cancel();
            socket = null;
        }

        Log.d(TAG, "Disconnected from socket");
    }

    public void reconnect() {
        if (attemptingReconnect) return;

        Log.d(TAG, String.format("Reconnecting to socket in %d ms", retryTime));

        disconnect();

        attemptingReconnect = true;

        // Exponential backoff
        SystemClock.sleep(retryTime);
        if (retryTime < RETRY_TIME_MAX) {
            retryTime *= 2;
        }

        connect();
    }

    public void update() {
        Log.d(TAG, "Requesting update from socket");

        if (socket == null) {
            connect();
            return;
        }

        socket.send("{ \"op\": 2 }");
    }

    @Override
    public void onOpen(WebSocket socket, Response response) {
        Log.d(TAG, "Socket connection opened");

        retryTime = RETRY_TIME_MIN;
        attemptingReconnect = false;

        // Handshake with socket
        final String authToken = authUtil.isAuthenticated() ? authUtil.getAuthTokenWithPrefix() : "";
        socket.send(String.format("{ \"op\": 0, \"d\": { \"auth\": \"%s\" } }", authToken));
    }

    @Override
    public void onMessage(WebSocket webSocket, String text) {
        Log.d(TAG, "Received message from socket: " + text);

        parseWebSocketResponse(text);
    }

    @Override
    public void onFailure(WebSocket webSocket, Throwable t, Response response) {
        Log.e(TAG, "Socket failure: " + t.getMessage(), t);
        reconnect();
    }

    @Override
    public void onClosed(WebSocket webSocket, int code, String reason) {
        Log.d(TAG, "Socket connection closed: " + reason);
        reconnect();
    }

    private void heartbeat(int milliseconds) {
        clearHeartbeat();

        heartbeatTask = new Runnable() {
            @Override
            public void run() {
                if (socket == null) return;

                Log.d(TAG, "Sending heartbeat to socket");
                socket.send("{ \"op\": 9 }");

                // Repeat
                heartbeatHandler.postDelayed(this, milliseconds);
            }
        };

        heartbeatHandler.postDelayed(heartbeatTask, milliseconds);
        Log.d(TAG, String.format("Created heartbeat task for %d ms", milliseconds));
    }

    private void clearHeartbeat() {
        if (heartbeatTask != null) {
            Log.d(TAG, "Removing heartbeat task");
            heartbeatHandler.removeCallbacksAndMessages(null);
            heartbeatTask = null;
        }
    }

    private void parseWebSocketResponse(String jsonString) {
        if (listener == null) {
            Log.d(TAG, "Listener is null");
            return;
        }

        if (jsonString == null) {
            listener.onSocketFailure();
            return;
        }

        try {
            final BaseResponse baseResponse = MOSHI.adapter(BaseResponse.class).fromJson(jsonString);
            switch (baseResponse.getOp()) {
                // Heartbeat init
                case 0:
                    final ConnectResponse connectResponse = MOSHI.adapter(ConnectResponse.class).fromJson(jsonString);
                    heartbeat(connectResponse.getD().getHeartbeat());
                    break;

                // Track update
                case 1:
                    final UpdateResponse updateResponse = MOSHI.adapter(UpdateResponse.class).fromJson(jsonString);
                    if (!isValidUpdate(updateResponse)) {
                        return;
                    }
                    listener.onSocketReceive(updateResponse.getD());
                    break;

                // Heartbeat ACK
                case 10:
                    break;

                default:
                    Log.d(TAG, "Received invalid socket data: " + jsonString);
            }
        } catch (IOException e) {
            Log.e(TAG, "Failed to parse socket data: " + jsonString, e);
        }
    }

    private boolean isValidUpdate(UpdateResponse updateResponse) {
        return updateResponse.getT().equals(TRACK_UPDATE)
                || updateResponse.getT().equals(TRACK_UPDATE_REQUEST)
                || updateResponse.getT().equals(QUEUE_UPDATE);
    }

    public interface Listener {
        void onSocketReceive(UpdateResponse.Details info);
        void onSocketFailure();
    }

}

File: app/src/main/java/me/echeung/moemoekyun/client/api/socket/Socket.kt
package me.echeung.moemoekyun.client.api.socket

import android.content.Context
import android.util.Log
import com.squareup.moshi.Moshi
import java.io.IOException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.ConflatedBroadcastChannel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import me.echeung.moemoekyun.client.RadioClient
import me.echeung.moemoekyun.client.api.socket.response.BaseResponse
import me.echeung.moemoekyun.client.api.socket.response.ConnectResponse
import me.echeung.moemoekyun.client.api.socket.response.EventNotificationResponse
import me.echeung.moemoekyun.client.api.socket.response.NotificationResponse
import me.echeung.moemoekyun.client.api.socket.response.UpdateResponse
import me.echeung.moemoekyun.client.network.NetworkClient
import me.echeung.moemoekyun.service.notification.EventNotification
import me.echeung.moemoekyun.util.ext.launchIO
import me.echeung.moemoekyun.util.ext.launchNow
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener

@OptIn(ExperimentalCoroutinesApi::class)
class Socket(
    private val context: Context,
    private val networkClient: NetworkClient
) : WebSocketListener() {

    val channel = ConflatedBroadcastChannel<SocketResult>()

    private val scope = CoroutineScope(Job() + Dispatchers.Main)

    private var retryTime = RETRY_TIME_MIN
    private var attemptingReconnect = false

    @Volatile
    private var socket: WebSocket? = null
    private val socketLock = Any()

    private var heartbeatJob: Job? = null

    fun connect() {
        synchronized(socketLock) {
            Log.d(TAG, "Connecting to socket...")

            if (socket != null) {
                disconnect()
            }

            val request = Request.Builder().url(RadioClient.library.socketUrl).build()
            socket = networkClient.client.newWebSocket(request, this)
        }
    }

    fun disconnect() {
        synchronized(socketLock) {
            clearHeartbeat()

            socket?.cancel()
            socket = null

            Log.d(TAG, "Disconnected from socket")
        }
    }

    fun reconnect() {
        if (attemptingReconnect) return

        Log.d(TAG, "Reconnecting to socket in $retryTime ms")

        disconnect()

        attemptingReconnect = true

        launchNow {
            delay(retryTime.toLong())

            // Exponential backoff
            if (retryTime < RETRY_TIME_MAX) {
                retryTime *= 2
            }

            connect()
        }
    }

    fun update() {
        synchronized(socketLock) {
            Log.d(TAG, "Requesting update from socket")

            if (socket == null) {
                connect()
                return
            }

            socket!!.send("{ \"op\": 2 }")
        }
    }

    override fun onOpen(webSocket: WebSocket, response: Response) {
        Log.d(TAG, "Socket connection opened")

        retryTime = RETRY_TIME_MIN
        attemptingReconnect = false
    }

    override fun onMessage(webSocket: WebSocket, text: String) {
        Log.d(TAG, "Received message from socket: $text")

        parseResponse(text)
    }

    override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
        Log.d(TAG, "Socket connection closed: $reason")

        reconnect()
    }

    override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
        Log.e(TAG, "Socket failure: ${t.message}", t)

        reconnect()
    }

    private fun initHeartbeat(delayMillis: Long) {
        clearHeartbeat()
        heartbeatJob = scope.launch {
            Log.d(TAG, "Created heartbeat job for $delayMillis ms")
            sendHeartbeat(delayMillis)
        }
    }

    private suspend fun sendHeartbeat(delayMillis: Long) {
        delay(delayMillis)

        if (heartbeatJob?.isActive != true || socket == null) {
            return
        }

        Log.d(TAG, "Sending heartbeat to socket")
        socket!!.send("{ \"op\": 9 }")

        // Repeat
        sendHeartbeat(delayMillis)
    }

    private fun clearHeartbeat() {
        Log.d(TAG, "Cancelling heartbeat job")
        heartbeatJob?.cancel()
    }

    private fun parseResponse(jsonString: String?) {
        if (jsonString == null) {
            launchIO {
                channel.send(SocketError())
            }
            return
        }

        try {
            val baseResponse = getResponse(BaseResponse::class.java, jsonString)
            when (baseResponse!!.op) {
                // Heartbeat init
                0 -> {
                    val connectResponse = getResponse(ConnectResponse::class.java, jsonString)
                    initHeartbeat(connectResponse!!.d!!.heartbeat.toLong())
                }

                // Update
                1 -> {
                    val updateResponse = getResponse(UpdateResponse::class.java, jsonString)
                    if (!isValidUpdate(updateResponse!!)) {
                        return
                    }
                    if (isNotification(updateResponse)) {
                        parseNotification(jsonString)
                        return
                    }

                    launchIO {
                        channel.send(SocketResponse(updateResponse.d))
                    }
                }

                // Heartbeat ACK
                10 -> {}

                else -> Log.d(TAG, "Received invalid socket data: $jsonString")
            }
        } catch (e: IOException) {
            Log.e(TAG, "Failed to parse socket data: $jsonString", e)
        }
    }

    private fun parseNotification(jsonString: String) {
        try {
            val notificationResponse = getResponse(NotificationResponse::class.java, jsonString)
            when (notificationResponse!!.t) {
                EventNotificationResponse.TYPE -> {
                    val eventResponse = getResponse(EventNotificationResponse::class.java, jsonString)
                    EventNotification.notify(context, eventResponse!!.d!!.event!!.name)
                }
            }
        } catch (e: IOException) {
            Log.e(TAG, "Failed to parse notification data: $jsonString", e)
        }
    }

    private fun isValidUpdate(updateResponse: UpdateResponse): Boolean {
        return (
            updateResponse.t == TRACK_UPDATE ||
                updateResponse.t == TRACK_UPDATE_REQUEST ||
                updateResponse.t == QUEUE_UPDATE ||
                isNotification(updateResponse)
            )
    }

    private fun isNotification(updateResponse: UpdateResponse): Boolean {
        return updateResponse.t == NOTIFICATION
    }

    @Throws(IOException::class)
    private fun <T : BaseResponse> getResponse(responseClass: Class<T>, jsonString: String): T? {
        return MOSHI.adapter(responseClass).fromJson(jsonString)
    }

    interface SocketResult
    class SocketResponse(val info: UpdateResponse.Details?) : SocketResult
    class SocketError : SocketResult

    companion object {
        private val TAG = Socket::class.java.simpleName

        private val MOSHI = Moshi.Builder().build()

        private const val TRACK_UPDATE = "TRACK_UPDATE"
        private const val TRACK_UPDATE_REQUEST = "TRACK_UPDATE_REQUEST"
        private const val QUEUE_UPDATE = "QUEUE_UPDATE"
        private const val NOTIFICATION = "NOTIFICATION"

        private const val RETRY_TIME_MIN = 250
        private const val RETRY_TIME_MAX = 4000
    }
}
